File: mini-agent-harness/typescript/agent/securityBoundary.ts
import path from 'node:path'

export type FilesystemPolicy = Readonly<{
  readRoots: readonly string[]
  writeRoots: readonly string[]
}>

export type NetworkPolicy = Readonly<{
  allowedHosts: readonly string[]
}>

export type ProcessRule = Readonly<{
  executable: string
  allowedArgvPrefixes: readonly (readonly string[])[]
}>

export type ExtensionTrustRule = Readonly<{
  extensionId: string
  source: string
  digest?: string
}>

export type SecurityPolicy = Readonly<{
  revision: number
  requireSandbox: boolean
  allowedWorkerIds: readonly string[]
  filesystem: FilesystemPolicy
  network: NetworkPolicy
  processRules: readonly ProcessRule[]
  trustedExtensions: readonly ExtensionTrustRule[]
}>

export type SecurityPolicyDraft = Omit<SecurityPolicy, 'revision'>

export type EffectRequest =
  | Readonly<{ kind: 'filesystem'; operation: 'read' | 'write'; path: string }>
  | Readonly<{ kind: 'network'; operation: 'connect'; host: string }>
  | Readonly<{
      kind: 'process'
      operation: 'spawn'
      executable: string
      argv: readonly string[]
    }>

export type ExtensionProvenance = Readonly<{
  extensionId: string
  source: string
  digest?: string
}>

/** This is the model/control-plane envelope. It never carries secret values. */
export type ExecutionRequest = Readonly<{
  callId: string
  workerId: string
  policyRevision: number
  permissionGranted: boolean
  effect: EffectRequest
  secretRefs?: readonly string[]
  extension?: ExtensionProvenance
}>

/** This envelope exists only inside the trusted execution boundary. */
export type TrustedExecutionEnvelope = Readonly<{
  callId: string
  workerId: string
  policyRevision: number
  effect: EffectRequest
  resolvedSecrets: ReadonlyMap<string, string>
  extension?: ExtensionProvenance
}>

export type WorkerResult = Readonly<{
  status: 'completed'
  outputRef?: string
}>

export type SecurityReason =
  | 'allowed'
  | 'permission_denied'
  | 'stale_policy'
  | 'worker_denied'
  | 'worker_mismatch'
  | 'sandbox_unavailable'
  | 'capability_denied'
  | 'extension_untrusted'
  | 'secret_missing'
  | 'cancelled'

export type SecurityReport = Readonly<{
  callId: string
  decision: 'allow' | 'deny'
  reason: SecurityReason
  policyRevision: number
  workerId: string
  capability: EffectRequest['kind']
  secretRefIds: readonly string[]
  extensionId?: string
  sandboxed: boolean
}>

export type SecurityExecutionResult =
  | Readonly<{ ok: true; report: SecurityReport; workerResult: WorkerResult }>
  | Readonly<{ ok: false; report: SecurityReport }>

export interface SecretResolver {
  resolve(ref: string): string | undefined | Promise<string | undefined>
}
/**
 * A port, not an OS sandbox implementation. Production adapters must enforce
 * the envelope at the process/container boundary.
 */
export interface SandboxPort {
  readonly workerId: string
  isAvailable(): boolean
  execute(envelope: TrustedExecutionEnvelope, signal: AbortSignal): Promise<WorkerResult>
}

export class StalePolicyRevisionError extends Error {}

export class PolicyEngine {
  #policy: SecurityPolicy

  constructor(initial: SecurityPolicy) {
    this.#policy = freezePolicy(initial)
  }

  snapshot(): SecurityPolicy {
    return this.#policy
  }

  replace(expectedRevision: number, draft: SecurityPolicyDraft): SecurityPolicy {
    if (expectedRevision !== this.#policy.revision) {
      throw new StalePolicyRevisionError(
        `stale policy revision ${expectedRevision}; current=${this.#policy.revision}`,
      )
    }
    this.#policy = freezePolicy({ ...structuredClone(draft), revision: expectedRevision + 1 })
    return this.#policy
  }
}

export class SecurityExecutor {
  readonly #policies: PolicyEngine
  readonly #sandbox: SandboxPort
  readonly #secrets: SecretResolver

  constructor(policies: PolicyEngine, sandbox: SandboxPort, secrets: SecretResolver) {
    this.#policies = policies
    this.#sandbox = sandbox
    this.#secrets = secrets
  }

  async execute(request: ExecutionRequest, signal: AbortSignal): Promise<SecurityExecutionResult> {
    const secretRefs = Object.freeze([...(request.secretRefs ?? [])])
    const firstPolicy = this.#policies.snapshot()
    const deny = (reason: Exclude<SecurityReason, 'allowed'>): SecurityExecutionResult =>
      Object.freeze({ ok: false, report: reportFor(request, firstPolicy, secretRefs, reason) })

    if (signal.aborted) return deny('cancelled')
    if (!request.permissionGranted) return deny('permission_denied')
    if (request.policyRevision !== firstPolicy.revision) return deny('stale_policy')
    if (!firstPolicy.allowedWorkerIds.includes(request.workerId)) return deny('worker_denied')
    if (this.#sandbox.workerId !== request.workerId) return deny('worker_mismatch')
    if (firstPolicy.requireSandbox && !this.#sandbox.isAvailable()) return deny('sandbox_unavailable')
    if (!allowsEffect(firstPolicy, request.effect)) return deny('capability_denied')
    if (request.extension && !allowsExtension(firstPolicy, request.extension)) {
      return deny('extension_untrusted')
    }

    const resolvedSecrets = new Map<string, string>()
    for (const ref of secretRefs) {
      if (signal.aborted) return deny('cancelled')
      const value = await this.#secrets.resolve(ref)
      if (signal.aborted) return deny('cancelled')
      if (value === undefined) return deny('secret_missing')
      resolvedSecrets.set(ref, value)
    }

    // Policy may refresh while a secret store or broker is being consulted.
    if (this.#policies.snapshot().revision !== firstPolicy.revision) return deny('stale_policy')
    if (signal.aborted) return deny('cancelled')
    if (firstPolicy.requireSandbox && !this.#sandbox.isAvailable()) return deny('sandbox_unavailable')

    const envelope: TrustedExecutionEnvelope = Object.freeze({
      callId: request.callId,
      workerId: request.workerId,
      policyRevision: firstPolicy.revision,
      effect: freezeEffect(request.effect),
      resolvedSecrets,
      ...(request.extension ? { extension: Object.freeze({ ...request.extension }) } : {}),
    })
    const workerResult = await this.#sandbox.execute(envelope, signal)
    return Object.freeze({
      ok: true,
      workerResult,
      report: reportFor(request, firstPolicy, secretRefs, 'allowed'),
    })
  }
}

function allowsEffect(policy: SecurityPolicy, effect: EffectRequest): boolean {
  switch (effect.kind) {
    case 'filesystem': {
      const roots = effect.operation === 'read'
        ? policy.filesystem.readRoots
        : policy.filesystem.writeRoots
      return roots.some(root => isWithin(root, effect.path))
    }
    case 'network':
      return policy.network.allowedHosts.some(pattern => hostMatches(pattern, effect.host))
    case 'process':
      return policy.processRules.some(rule =>
        rule.executable === effect.executable
        && rule.allowedArgvPrefixes.some(prefix => startsWith(effect.argv, prefix)),
      )
  }
}

function allowsExtension(policy: SecurityPolicy, extension: ExtensionProvenance): boolean {
  return policy.trustedExtensions.some(rule =>
    rule.extensionId === extension.extensionId
    && rule.source === extension.source
    && (rule.digest === undefined || rule.digest === extension.digest),
  )
}

function isWithin(root: string, candidate: string): boolean {
  const relative = path.relative(path.resolve(root), path.resolve(candidate))
  return relative === '' || (!relative.startsWith('..') && !path.isAbsolute(relative))
}

function hostMatches(pattern: string, host: string): boolean {
  const normalizedPattern = pattern.toLowerCase()
  const normalizedHost = host.toLowerCase()
  return normalizedPattern.startsWith('*.')
    ? normalizedHost.endsWith(normalizedPattern.slice(1))
    : normalizedHost === normalizedPattern
}

function startsWith(actual: readonly string[], prefix: readonly string[]): boolean {
  return prefix.length <= actual.length && prefix.every((value, index) => actual[index] === value)
}

function reportFor(
  request: ExecutionRequest,
  policy: SecurityPolicy,
  secretRefs: readonly string[],
  reason: SecurityReason,
): SecurityReport {
  return Object.freeze({
    callId: request.callId,
    decision: reason === 'allowed' ? 'allow' : 'deny',
    reason,
    policyRevision: policy.revision,
    workerId: request.workerId,
    capability: request.effect.kind,
    secretRefIds: secretRefs,
    ...(request.extension ? { extensionId: request.extension.extensionId } : {}),
    sandboxed: policy.requireSandbox && reason === 'allowed',
  })
}

function freezePolicy(policy: SecurityPolicy): SecurityPolicy {
  if (!Number.isInteger(policy.revision) || policy.revision < 1) {
    throw new Error('policy revision must be a positive integer')
  }
  if (policy.allowedWorkerIds.length === 0) throw new Error('policy requires at least one worker')
  return deepFreeze(structuredClone(policy))
}

function freezeEffect(effect: EffectRequest): EffectRequest {
  return deepFreeze(structuredClone(effect))
}

function deepFreeze<T>(value: T): T {
  if (value && typeof value === 'object' && !Object.isFrozen(value)) {
    Object.freeze(value)
    for (const child of Object.values(value)) deepFreeze(child)
  }
  return value
}

File: mini-agent-harness/python/security_boundary.py
from __future__ import annotations

import asyncio
import os
from dataclasses import dataclass, replace
from pathlib import Path
from types import MappingProxyType
from typing import Awaitable, Callable, Literal, Mapping, Protocol


@dataclass(frozen=True)
class ProcessRule:
    executable: str
    allowed_argv_prefixes: tuple[tuple[str, ...], ...]


@dataclass(frozen=True)
class ExtensionTrustRule:
    extension_id: str
    source: str
    digest: str | None = None


@dataclass(frozen=True)
class SecurityPolicy:
    revision: int
    require_sandbox: bool
    allowed_worker_ids: tuple[str, ...]
    read_roots: tuple[str, ...]
    write_roots: tuple[str, ...]
    allowed_hosts: tuple[str, ...]
    process_rules: tuple[ProcessRule, ...]
    trusted_extensions: tuple[ExtensionTrustRule, ...]


@dataclass(frozen=True)
class EffectRequest:
    kind: Literal["filesystem", "network", "process"]
    operation: Literal["read", "write", "connect", "spawn"]
    target: str
    argv: tuple[str, ...] = ()


@dataclass(frozen=True)
class ExtensionProvenance:
    extension_id: str
    source: str
    digest: str | None = None


@dataclass(frozen=True)
class ExecutionRequest:
    call_id: str
    worker_id: str
    policy_revision: int
    permission_granted: bool
    effect: EffectRequest
    secret_refs: tuple[str, ...] = ()
    extension: ExtensionProvenance | None = None


@dataclass(frozen=True)
class TrustedExecutionEnvelope:
    call_id: str
    worker_id: str
    policy_revision: int
    effect: EffectRequest
    resolved_secrets: Mapping[str, str]
    extension: ExtensionProvenance | None = None


@dataclass(frozen=True)
class WorkerResult:
    status: Literal["completed"] = "completed"
    output_ref: str | None = None


SecurityReason = Literal[
    "allowed",
    "permission_denied",
    "stale_policy",
    "worker_denied",
    "worker_mismatch",
    "sandbox_unavailable",
    "capability_denied",
    "extension_untrusted",
    "secret_missing",
    "cancelled",
]


@dataclass(frozen=True)
class SecurityReport:
    call_id: str
    decision: Literal["allow", "deny"]
    reason: SecurityReason
    policy_revision: int
    worker_id: str
    capability: Literal["filesystem", "network", "process"]
    secret_ref_ids: tuple[str, ...]
    extension_id: str | None
    sandboxed: bool


@dataclass(frozen=True)
class SecurityExecutionResult:
    ok: bool
    report: SecurityReport
    worker_result: WorkerResult | None = None


class SandboxPort(Protocol):
    worker_id: str

    def is_available(self) -> bool: ...

    async def execute(
        self, envelope: TrustedExecutionEnvelope, signal: "CancellationSignal"
    ) -> WorkerResult: ...


class CancellationSignal:
    def __init__(self) -> None:
        self.cancelled = False
        self.reason = "cancelled"

    def cancel(self, reason: str = "cancelled") -> None:
        self.cancelled = True
        self.reason = reason

    def throw_if_cancelled(self) -> None:
        if self.cancelled:
            raise asyncio.CancelledError(self.reason)


class StalePolicyRevisionError(ValueError):
    pass


class PolicyEngine:
    def __init__(self, initial: SecurityPolicy) -> None:
        self._policy = _validate_policy(initial)

    def snapshot(self) -> SecurityPolicy:
        return self._policy

    def replace(
        self,
        expected_revision: int,
        draft: SecurityPolicy,
    ) -> SecurityPolicy:
        if expected_revision != self._policy.revision:
            raise StalePolicyRevisionError(
                f"stale policy revision {expected_revision}; current={self._policy.revision}"
            )
        self._policy = _validate_policy(
            replace(draft, revision=expected_revision + 1)
        )
        return self._policy


SecretResolver = Callable[[str], str | None | Awaitable[str | None]]


class SecurityExecutor:
    def __init__(
        self,
        policies: PolicyEngine,
        sandbox: SandboxPort,
        secrets: SecretResolver,
    ) -> None:
        self._policies = policies
        self._sandbox = sandbox
        self._secrets = secrets

    async def execute(
        self, request: ExecutionRequest, signal: CancellationSignal
    ) -> SecurityExecutionResult:
        policy = self._policies.snapshot()

        def deny(reason: SecurityReason) -> SecurityExecutionResult:
            return SecurityExecutionResult(False, _report(request, policy, reason))

        if signal.cancelled:
            return deny("cancelled")
        if not request.permission_granted:
            return deny("permission_denied")
        if request.policy_revision != policy.revision:
            return deny("stale_policy")
        if request.worker_id not in policy.allowed_worker_ids:
            return deny("worker_denied")
        if self._sandbox.worker_id != request.worker_id:
            return deny("worker_mismatch")
        if policy.require_sandbox and not self._sandbox.is_available():
            return deny("sandbox_unavailable")
        if not _allows_effect(policy, request.effect):
            return deny("capability_denied")
        if request.extension and not _allows_extension(policy, request.extension):
            return deny("extension_untrusted")

        resolved: dict[str, str] = {}
        for ref in request.secret_refs:
            if signal.cancelled:
                return deny("cancelled")
            value = self._secrets(ref)
            if isinstance(value, Awaitable):
                value = await value
            if signal.cancelled:
                return deny("cancelled")
            if value is None:
                return deny("secret_missing")
            resolved[ref] = value

        if self._policies.snapshot().revision != policy.revision:
            return deny("stale_policy")
        if signal.cancelled:
            return deny("cancelled")
        if policy.require_sandbox and not self._sandbox.is_available():
            return deny("sandbox_unavailable")

        envelope = TrustedExecutionEnvelope(
            request.call_id,
            request.worker_id,
            policy.revision,
            request.effect,
            MappingProxyType(resolved),
            request.extension,
        )
        worker_result = await self._sandbox.execute(envelope, signal)
        return SecurityExecutionResult(
            True, _report(request, policy, "allowed"), worker_result
        )


def _allows_effect(policy: SecurityPolicy, effect: EffectRequest) -> bool:
    if effect.kind == "filesystem":
        roots = policy.read_roots if effect.operation == "read" else policy.write_roots
        return any(_is_within(root, effect.target) for root in roots)
    if effect.kind == "network":
        return any(_host_matches(pattern, effect.target) for pattern in policy.allowed_hosts)
    return any(
        rule.executable == effect.target
        and any(effect.argv[: len(prefix)] == prefix for prefix in rule.allowed_argv_prefixes)
        for rule in policy.process_rules
    )


def _allows_extension(
    policy: SecurityPolicy, extension: ExtensionProvenance
) -> bool:
    return any(
        rule.extension_id == extension.extension_id
        and rule.source == extension.source
        and (rule.digest is None or rule.digest == extension.digest)
        for rule in policy.trusted_extensions
    )


def _is_within(root: str, candidate: str) -> bool:
    try:
        Path(candidate).resolve().relative_to(Path(root).resolve())
        return True
    except ValueError:
        return False


def _host_matches(pattern: str, host: str) -> bool:
    pattern = pattern.lower()
    host = host.lower()
    return host.endswith(pattern[1:]) if pattern.startswith("*.") else host == pattern


def _report(
    request: ExecutionRequest, policy: SecurityPolicy, reason: SecurityReason
) -> SecurityReport:
    return SecurityReport(
        request.call_id,
        "allow" if reason == "allowed" else "deny",
        reason,
        policy.revision,
        request.worker_id,
        request.effect.kind,
        tuple(request.secret_refs),
        request.extension.extension_id if request.extension else None,
        policy.require_sandbox and reason == "allowed",
    )


def _validate_policy(policy: SecurityPolicy) -> SecurityPolicy:
    if policy.revision < 1:
        raise ValueError("policy revision must be positive")
    if not policy.allowed_worker_ids:
        raise ValueError("policy requires at least one worker")
    return policy
